src/java/org/apache/cassandra/schema/TableMetadata.java

```diff
@@ -1342,6 +1343,66 @@ private void appendDropColumns(CqlBuilder builder)
         }
     }
 
+    /**
+     * Returns a string representation of a partition in a CQL-friendly format.
+     *
+     * For non-composite types it returns the result of {@link org.apache.cassandra.cql3.CQL3Type#toCQLLiteral}
+     * applied to the partition key.
+     * For composite types it applies {@link org.apache.cassandra.cql3.CQL3Type#toCQLLiteral} to each subkey and
+     * combines the results into a tuple.
+     *
+     * @param partitionKey a partition key
+     * @return CQL-like string representation of a partition key
+     */
+    public String partitionKeyAsCQLLiteral(ByteBuffer partitionKey)
+    {
+        return primaryKeyAsCQLLiteral(partitionKey, Clustering.EMPTY);
+    }
+
+    /**
+     * Returns a string representation of a primary key in a CQL-friendly format.
+     *
+     * @param partitionKey the partition key part of the primary key
+     * @param clustering the clustering key part of the primary key
+     * @return a CQL-like string representation of the specified primary key
+     */
+    public String primaryKeyAsCQLLiteral(ByteBuffer partitionKey, Clustering<?> clustering)
+    {
+        int clusteringSize = clustering.size();
+
+        String[] literals;
+        int i = 0;
+
+        if (partitionKeyType instanceof CompositeType)
+        {
+            List<AbstractType<?>> components = partitionKeyType.getComponents();
+            int size = components.size();
+            literals = new String[size + clusteringSize];
+            ByteBuffer[] values = ((CompositeType) partitionKeyType).split(partitionKey);
+            for (i = 0; i < size; i++)
+            {
+                literals[i] = asCQLLiteral(components.get(i), values[i]);
+            }
+        }
+        else
+        {
+            literals = new String[1 + clusteringSize];
+            literals[i++] = asCQLLiteral(partitionKeyType, partitionKey);
+        }
+
+        for (int j = 0; j < clusteringSize; j++)
+        {
+            literals[i++] = asCQLLiteral(clusteringColumns().get(j).type, clustering.bufferAt(j));
+        }
+
+        return i == 1 ? literals[0] : "(" + String.join(", ", literals) + ")";
```

Comment:
Don't we need more `()` in case of composites of the form `((race_year, race_name), rank)`?

Reply:
Do you mean for the partition key part inside the clustering key, like when we define the primary key while creating a table? We only use that syntax when defining tables, but not when writing actual data to it:
```
CREATE TABLE t (k1 int, k2 int, c1 int, c2 int, PRIMARY KEY ((k1, k2), c1, c2));
INSERT INTO t (k1, k2, c1, c2) VALUES (1, 2, 3, 4);
```
Since here we are printing the data itself and not the column names I think that using a single set of parenthesis looks a bit more aligned with CQL syntax. In the example above, this method would return `(1, 2, 3 , 4)` rather than `((1, 2), 3, 4)`.

Reply:
Ok I'm fine with it.